examples/matrix_multiplication.masm

```diff
@@ -0,0 +1,735 @@
+end
+
+# Multiply two matrices `A*B=C`.
+#
+# A must be in row-major order
+# B must be in column-major order
+# C will be in row-major order
+#
+# Notes: 
+# - `addrC` must different to `addrA` and `addrB`
+# - This procedure only works with matrixes where both row and col are u32
+# values
+proc.matrix_multiply.0
+    # input: [addrA, addrB, addrC, ...]
+    # output: [...]
+
+    # copy dimensions of A
+    # [[row1, col1, 0, 0], addrA, addrB, addrC, ...]
+    padw dup.4 mem_loadw
+
+    # copy dimensions of B
+    # [[row2, col2, 0, 0], [row1, col1, 0, 0], addrA, addrB, addrC, ...]
+    padw dup.9 mem_loadw
+
+    # assert row2==col1 requirement for matrix multiplication
+    # [[col2, 0, 0, row1], [col1, 0, 0, addrA], addrB, addrC, ...]
+    dup.5 assert_eq
+
+    # Set C's size, where row=row1 col=col2
+    # [[row, col, 0, 0], [col1, 0, 0, addrA], addrB, addrC, ...]
+    movup.3 dup.9 mem_storew
+
+    # count the number of words the inner most loop have to consume.
+    # [[words, row, col, 0], [0, 0, 0, addrA], addrB, addrC, ...]
+    movup.4 exec.count_to_word
+
+    # discard padding in the middle of the data
+    # [[words, row, col, 0], addrA, addrB, addrC, ...]
+    push.0 movdn.4 swapw dropw
+
+    # udpate all indices by one, accounts for the size being set
+    # [[words, row, col, 0], addrA+1, addrB+1, addrC+1, ...]
+    swapw add.1 swap add.1 movup.2 add.1 swap.2 swapw
+
+    # [[0, row, col, addrC+1], [0, addrA+1, addrB+1, words, ...]
+    swap.6 movdn.2 push.0
+
+    # Order elements as [words, addrA+1, addrB+1, 0]. This reduces the number of
+    # maintenance operations needed in the inner loop. (values are moved only
+    # once, from back to front, when an update is necessary)
+    #
+    # [[0, row, col, addrC+1], [words, addrA+1, addrB+1, 0], ...]
+    swapw movdn.2 movup.3 swapw
+
+    # `r`: starts at `row` and decrements until `0`
+    # `writePtr`: starts at `addrC+1` and increment as data is written
+    # `readA`: starts at `addrA+1` and increments by `words` after every row is
+    # computed
+    # [[r, 0, col, writePtr], [words, readA, addrB, 0], ...]
+    swap
+
+    ## # DEBUG
+    ## dup.0 eq.2 assert   # rows(A)
+    ## dup.1 eq.0 assert   # always 0
+    ## dup.2 eq.2 assert   # columns(B)
+    ## dup.3 eq.301 assert # address(C) + 1
+    ## dup.4 eq.1 assert   # round_up(columns(A) / 4)
+    ## dup.5 eq.101 assert # address(A) + 1
+    ## dup.6 eq.201 assert # address(B) + 1
+    ## dup.7 eq.0 assert
+    ## # /DEBUG
+
+    # ROW: [[r, 0, col, writePtr], [words, readA, addrB, 0], ...]
+    # while(r!=0)
+    dup neq.0 while.true
+
+        # The loop counter `c` will be increasing from 0 to col.
+        # Used for:
+        #   1. End condition when c==col
+        #   2. Offset into addrB
+        #   3. To determine how many padding zeros are needed for row
+        # [[c, r, col, writePtr], [words, readA, addrB, 0], ...]
+        swap
+
+        ## # DEBUG
+        ## dup.0 eq.0 assert         # c==0
```

Comment:
I left debug statements as a sample strategy for loops. It is a bit hard to see what is going on in a loop, specially if the first iteration is successful and a later one fails. In that case testing for code invariants seems to be the best approach.

Initially I started with a `dupw` here, but that made it tricky to understand the number of cycles the assertions would take. This is a problem because at the current version the only reference we have is the number reported in `FailedAssertion`, which makes it hard to figure out which of the assertions failed. This pattern allows all other invariants to be commented out to help figure out what went wrong.

Reply:
> In that case testing for code invariants seems to be the best approach.

Agreed. It's a useful pattern - clear, logical, readable.